File: ios/RaiseTimer/Domain/TimerEngine.swift
import Foundation

/// Pure transformations on `TournamentState`. Driven by wall-clock deltas supplied by the
/// caller so it survives backgrounding and app restarts. Mirrors the Android `TimerEngine`.
enum TimerEngine {
    struct TickResult {
        var state: TournamentState
        var levelAdvanced: Bool
        var finished: Bool
    }

    /// Apply wall-clock delta between `state.lastTickAt` and `now`. Handles normal 1-second
    /// foreground ticks as well as multi-second/multi-level catch-up after a background or
    /// cold start. Sub-second remainder is preserved in the returned `lastTickAt`.
    static func catchUp(_ state: TournamentState, now: Double) -> TickResult {
        guard state.isRunning else {
            return TickResult(state: state, levelAdvanced: false, finished: false)
        }
        guard let anchor = state.lastTickAt else {
            var copy = state
            copy.lastTickAt = now
            return TickResult(state: copy, levelAdvanced: false, finished: false)
        }

        let deltaSeconds = Int((now - anchor))
        if deltaSeconds <= 0 {
            return TickResult(state: state, levelAdvanced: false, finished: false)
        }

        var budget = deltaSeconds
        var currentIndex = state.currentLevelIndex
        var remaining = state.remainingSeconds
        var advanced = false
        var finished = false

        while budget > 0 {
            if remaining > budget {
                remaining -= budget
                budget = 0
            } else {
                budget -= remaining
                remaining = 0
                let nextIndex = currentIndex + 1
                if nextIndex > state.config.levels.count - 1 {
                    advanced = true
                    finished = true
                    break
                }
                currentIndex = nextIndex
                remaining = state.config.levels[nextIndex].durationSeconds
                advanced = true
            }
        }

        var copy = state
        copy.currentLevelIndex = currentIndex
        copy.remainingSeconds = remaining
        copy.isRunning = !finished
        copy.lastTickAt = finished ? nil : anchor + Double(deltaSeconds)
        return TickResult(state: copy, levelAdvanced: advanced, finished: finished)
    }

    static func start(_ state: TournamentState, now: Double) -> TournamentState {
        var copy = state
        copy.isRunning = true
        copy.lastTickAt = now
        return copy
    }

    static func pause(_ state: TournamentState) -> TournamentState {
        var copy = state
        copy.isRunning = false
        copy.lastTickAt = nil
        return copy
    }

    static func reset(_ state: TournamentState) -> TournamentState {
        var copy = state
        copy.currentLevelIndex = 0
        copy.remainingSeconds = state.config.levels.first?.durationSeconds ?? 0
        copy.isRunning = false
        copy.lastTickAt = nil
        return copy
    }

    static func nextLevel(_ state: TournamentState, now: Double) -> TournamentState {
        var copy = state
        let clamped = min(state.currentLevelIndex + 1, state.config.levels.count - 1)
        copy.currentLevelIndex = clamped
        copy.remainingSeconds = state.config.levels[clamped].durationSeconds
        // If running, reset the tick anchor so the fresh level starts counting from now.
        copy.lastTickAt = state.isRunning ? now : nil
        return copy
    }

    static func previousLevel(_ state: TournamentState, now: Double) -> TournamentState {
        var copy = state
        let clamped = max(state.currentLevelIndex - 1, 0)
        copy.currentLevelIndex = clamped
        copy.remainingSeconds = state.config.levels[clamped].durationSeconds
        copy.lastTickAt = state.isRunning ? now : nil
        return copy
    }
}

File: android/app/src/main/java/com/jiny/raisetimer/domain/TimerEngine.kt
package com.jiny.raisetimer.domain

import com.jiny.raisetimer.domain.model.TournamentState

/**
 * Pure, stateless transformations on [TournamentState]. The engine is driven entirely by
 * wall-clock deltas supplied by the caller so it survives backgrounding and process restarts:
 * the UI layer calls [catchUp] on every tick and on app resume, passing the current epoch
 * millis. No internal time source — keeps the engine unit-testable and side-effect free.
 */
object TimerEngine {

    data class TickResult(
        val state: TournamentState,
        val levelAdvanced: Boolean,
        val finished: Boolean,
    )

    /**
     * Apply wall-clock delta between [TournamentState.lastTickAt] and [nowMillis]. Handles
     * the common 1-second foreground tick as well as multi-second/multi-level catch-up after
     * a background or cold start. Sub-second remainder is preserved in the new [lastTickAt]
     * so repeated calls stay phase-aligned.
     */
    fun catchUp(state: TournamentState, nowMillis: Long): TickResult {
        if (!state.isRunning) return TickResult(state, levelAdvanced = false, finished = false)
        val anchor = state.lastTickAt ?: return TickResult(
            state.copy(lastTickAt = nowMillis),
            levelAdvanced = false,
            finished = false,
        )

        val deltaSeconds = ((nowMillis - anchor) / 1000L).toInt()
        if (deltaSeconds <= 0) return TickResult(state, levelAdvanced = false, finished = false)

        var budget = deltaSeconds
        var currentIndex = state.currentLevelIndex
        var remaining = state.remainingSeconds
        var advanced = false
        var finished = false

        while (budget > 0) {
            if (remaining > budget) {
                remaining -= budget
                budget = 0
            } else {
                budget -= remaining
                remaining = 0
                val nextIndex = currentIndex + 1
                if (nextIndex > state.config.levels.lastIndex) {
                    advanced = true
                    finished = true
                    break
                }
                currentIndex = nextIndex
                remaining = state.config.levels[nextIndex].durationSeconds
                advanced = true
            }
        }

        val consumedMillis = deltaSeconds * 1000L
        val newState = state.copy(
            currentLevelIndex = currentIndex,
            remainingSeconds = remaining,
            isRunning = !finished,
            lastTickAt = if (finished) null else anchor + consumedMillis,
        )
        return TickResult(newState, levelAdvanced = advanced, finished = finished)
    }

    fun start(state: TournamentState, nowMillis: Long): TournamentState =
        state.copy(isRunning = true, lastTickAt = nowMillis)

    fun pause(state: TournamentState): TournamentState =
        state.copy(isRunning = false, lastTickAt = null)

    fun reset(state: TournamentState): TournamentState = state.copy(
        currentLevelIndex = 0,
        remainingSeconds = state.config.levels.first().durationSeconds,
        isRunning = false,
        lastTickAt = null,
    )

    fun nextLevel(state: TournamentState, nowMillis: Long): TournamentState {
        val nextIndex = (state.currentLevelIndex + 1).coerceAtMost(state.config.levels.lastIndex)
        return state.copy(
            currentLevelIndex = nextIndex,
            remainingSeconds = state.config.levels[nextIndex].durationSeconds,
            // If running, reset the tick anchor so the fresh level starts counting from now.
            lastTickAt = if (state.isRunning) nowMillis else null,
        )
    }

    fun previousLevel(state: TournamentState, nowMillis: Long): TournamentState {
        val prevIndex = (state.currentLevelIndex - 1).coerceAtLeast(0)
        return state.copy(
            currentLevelIndex = prevIndex,
            remainingSeconds = state.config.levels[prevIndex].durationSeconds,
            lastTickAt = if (state.isRunning) nowMillis else null,
        )
    }
}
